### ISSUE TITLE
grep2 does not work on Windows/Vim

### ISSUE THREAD

**Issue by Grueslayer** [liuchengxu/vim-clap#533]
- OS: MS-Windows 10 
 - Vim: 8.2.1079
 - vim-clap: 0.1.21
 - Have you reproduced with a minimal vimrc: yes
 - Have you updated to the latest plugin version: yes
 - Have you upgraded to/compiled the latest Rust binary: yes

1. grep2 shows window but entering any search pattern does nothing

Reproduce:

```vim
:cd c:\dev
:Clap grep2
```

Enter `Sana` (or anything you want)

Wait 3 minutes, nothing happens. `:messages` keeps clean

2. CR leads to an error

```
vim-clap: target_dir:C:\Dev, Run:function('116', {'is_sync': function('135'), 'get_source': function('134'), 'init_default_impl': function('138'), 'can_async': function('137'), 'init_display_win': function('139'), 'support_multi_select': function('126'), '_apply_sink': function('116'), 'id': 'grep2', 'filter': function('125'), 'sink_star': function('119'), 'sink': function('118'), 'support_open_action': function('127'), '_': function('114'), 'is_pure_async': function('136'), 'try_set_syntax': function('129'), 'on_typed': function('121'), 'source_async_or_default': function('131'), 'on_exit': function('123'), 'has_enable_rooter': function('117'), 'source_async': function('132'), 'jobstop': function('124'), 'abort': function('115'), 'is_rpc_type': function('128'), 'apply_query': function('130'), 'on_enter': function('120'), '_apply_source': function('133'), 'on_move': function('122')}), run_args:[''], exception:Vim(let):E684: list index out of range: 1
```

Reproduce:

Doing everything from 1. and press Enter/Return

3. Pressing BS until an empty search index gives following error

```
vim-clap: error: The following required arguments were not provided:
vim-clap:     <grep-query>
vim-clap: 
vim-clap: USAGE:
vim-clap:     maple.exe grep <grep-query> --cmd-dir <cmd-dir>
vim-clap: 
vim-clap: For more information try --help
```

Reproduce:

Doing everything from 1. pressing 4 time Backspace (removing `Sana`)

**Comment by liuchengxu** [comment]
You should fill the issue template and provide more env info to help others investigate the issue.

**Comment by Grueslayer** [comment]
It seems there is a problem with calling rg

```
rg  --column --line-number --no-heading --color=never --smart-case ''
```

1. It is called without my search string (maybe that's right because Clap will filter that later, but that will be really slow if rg.ex gives you the complete tree)

2. the directory is not given to rg,exe by commandline and it seems that the process working directory is not set to Vims one. So it looks like "rg" is scanning my complete filesystem (will take hours)

**Comment by liuchengxu** [comment]
If so, you can quickly apply this patch and then recompile the binary to see if it helps. #60 is related.

```diff
diff --git a/crates/maple_cli/src/cmd/grep.rs b/crates/maple_cli/src/cmd/grep.rs
index 02ccfc8..548dd2a 100644
--- a/crates/maple_cli/src/cmd/grep.rs
+++ b/crates/maple_cli/src/cmd/grep.rs
@@ -18,7 +18,7 @@ const RG_ARGS: [&str; 7] = [
     "",
 ];

-const RG_EXEC_CMD: &str = "rg --column --line-number --no-heading --color=never --smart-case ''";
+const RG_EXEC_CMD: &str = "rg --column --line-number --no-heading --color=never --smart-case '' .";

 #[derive(StructOpt, Debug, Clone)]
 pub struct Grep {
```

**Comment by Grueslayer** [comment]
Okay... that fixes Point 1 👍 

Point 2 still exists (when list is empty, e.g. starting and not enterng any text)
Point 3 still fails

**Comment by liuchengxu** [comment]
Well, good to see some progress here, but I have to say it's not wise to list multiple glitches in one issue, so much easier to forget  -).

**Comment by Grueslayer** [comment]
Maybe we can both go further with 3.
This is the part handling the input in grep2:

```vim
function! s:grep2.on_typed()
  if exists('g:__clap_forerunner_tempfile')
    call clap#filter#async#dyn#grep_from_cache(g:__clap_forerunner_tempfile)
  elseif exists('g:__clap_forerunner_result')
    let query = g:clap.input.get()
    if query ==# ''
      return
    endif
    call clap#filter#on_typed(function('clap#filter#sync'), query, g:__clap_forerunner_result)
  else
    call clap#filter#async#dyn#start_grep()
  endif
endfunction    
```

At least on windows `on_complete_maple` is not called and therefore __clap_forerunner_tempfile nore _result is ever set. We always call start_grep. And that does not work for an empty string.

**Comment by Grueslayer** [comment]
> 
> 
> If so, you can quickly apply this patch and then recompile the binary to see if it helps. #60 is related.
> 
> ```diff
> diff --git a/crates/maple_cli/src/cmd/grep.rs b/crates/maple_cli/src/cmd/grep.rs
> index 02ccfc8..548dd2a 100644
> --- a/crates/maple_cli/src/cmd/grep.rs
> +++ b/crates/maple_cli/src/cmd/grep.rs
> @@ -18,7 +18,7 @@ const RG_ARGS: [&str; 7] = [
>      "",
>  ];
> 
> -const RG_EXEC_CMD: &str = "rg --column --line-number --no-heading --color=never --smart-case ''";
> +const RG_EXEC_CMD: &str = "rg --column --line-number --no-heading --color=never --smart-case '' .";
> 
>  #[derive(StructOpt, Debug, Clone)]
>  pub struct Grep {
> ```

Can you please apply that to the repo to fix at least one part of this ticket?

**Comment by liuchengxu** [comment]
@Grueslayer Thanks for the reminder, applied for the windows.

**Comment by Grueslayer** [comment]
That's also needed on MacOS, I think that will work on Linux, too.

**Comment by liuchengxu** [comment]
I use macOS/Linux and have no issue without it.

**Comment by Grueslayer** [comment]
I mainly use MacOS and have the same issue there and this fixed it, too.

**Comment by liuchengxu** [comment]
Does `rg --column --line-number --no-heading --color=never --smart-case ''` work in the command line?

**Comment by Grueslayer** [comment]
Of course! You can also call that on Windows without any problems!

This happens using it WITHOUT the fix on MacOS and MacVim

```
 UID   PID  PPID   C STIME   TTY           TIME CMD
  501  3298  3297   0  7:16pm ??         0:00.01 rg --column --line-number --no-heading --color=never --smart-case
```

```
PID   COMMAND      %CPU TIME     #TH   #WQ  #PORT MEM    PURG   CMPR PGRP PPID STATE
ng
3298  rg           0.0  00:00.00 1     0    10    1468K  0B     0B   3297 3297 sleeping
```

You see it is sleeping, soing nothing.

Maybe thats not an OS, maybe thats a nvim/vim difference in process environment.

BUT it does not hurt anyone explicitly giving "." as current directory.

**Comment by liuchengxu** [comment]
All problems should be addressed in #712 now.
